File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/17-分账/QueryReturnOrder.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.HashMap;
import java.util.Map;

/**
 * 查询分账回退结果（服务商模式）
 */
public class QueryReturnOrder {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "GET";
  private static String PATH = "/v3/profitsharing/return-orders/{out_return_no}";

  public static void main(String[] args) {
    QueryReturnOrder client = new QueryReturnOrder(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    QueryReturnOrderRequest request = new QueryReturnOrderRequest();
    request.outReturnNo = "R20190516001";
    request.subMchid = "1900000109";
    request.outOrderNo = "P20190806125346";
    try {
      ReturnOrdersEntity response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public ReturnOrdersEntity run(QueryReturnOrderRequest request) {
    String uri = PATH;
    uri = uri.replace("{out_return_no}", WXPayUtility.urlEncode(request.outReturnNo));
    Map<String, Object> args = new HashMap<>();
    args.put("sub_mchid", request.subMchid);
    args.put("out_order_no", request.outOrderNo);
    String queryString = WXPayUtility.urlEncode(args);
    if (!queryString.isEmpty()) {
      uri = uri + "?" + queryString;
    }

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, null));
    reqBuilder.method(METHOD, null);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, ReturnOrdersEntity.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public QueryReturnOrder(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class QueryReturnOrderRequest {
    @SerializedName("out_return_no")
    @Expose(serialize = false)
    public String outReturnNo;

    @SerializedName("sub_mchid")
    @Expose(serialize = false)
    public String subMchid;

    @SerializedName("out_order_no")
    @Expose(serialize = false)
    public String outOrderNo;
  }

  public static class ReturnOrdersEntity {
    @SerializedName("sub_mchid")
    public String subMchid;

    @SerializedName("order_id")
    public String orderId;

    @SerializedName("out_order_no")
    public String outOrderNo;

    @SerializedName("out_return_no")
    public String outReturnNo;

    @SerializedName("return_id")
    public String returnId;

    @SerializedName("return_mchid")
    public String returnMchid;

    @SerializedName("amount")
    public Long amount;

    @SerializedName("description")
    public String description;

    @SerializedName("result")
    public ReturnOrderStatus result;

    @SerializedName("fail_reason")
    public ReturnOrderFailReason failReason;

    @SerializedName("create_time")
    public String createTime;

    @SerializedName("finish_time")
    public String finishTime;
  }

  public enum ReturnOrderStatus {
    @SerializedName("PROCESSING")
    PROCESSING,
    @SerializedName("SUCCESS")
    SUCCESS,
    @SerializedName("FAILED")
    FAILED
  }

  public enum ReturnOrderFailReason {
    @SerializedName("ACCOUNT_ABNORMAL")
    ACCOUNT_ABNORMAL,
    @SerializedName("BALANCE_NOT_ENOUGH")
    BALANCE_NOT_ENOUGH,
    @SerializedName("TIME_OUT_CLOSED")
    TIME_OUT_CLOSED,
    @SerializedName("PAYER_ACCOUNT_ABNORMAL")
    PAYER_ACCOUNT_ABNORMAL,
    @SerializedName("INVALID_REQUEST")
    INVALID_REQUEST
  }

}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/17-分账/QueryReturnOrder.go
package main

import (
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &QueryReturnOrderRequest{
		OutReturnNo: wxpay_utility.String("R20190516001"),
		SubMchid:    wxpay_utility.String("1900000109"),
		OutOrderNo:  wxpay_utility.String("P20190806125346"),
	}

	response, err := QueryReturnOrder(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}

	fmt.Printf("请求成功: %+v\n", response)
}

func QueryReturnOrder(config *wxpay_utility.MchConfig, request *QueryReturnOrderRequest) (response *ReturnOrdersEntity, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "GET"
		path   = "/v3/profitsharing/return-orders/{out_return_no}"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqUrl.Path = strings.Replace(reqUrl.Path, "{out_return_no}", url.PathEscape(*request.OutReturnNo), -1)
	query := reqUrl.Query()
	if request.SubMchid != nil {
		query.Add("sub_mchid", *request.SubMchid)
	}
	if request.OutOrderNo != nil {
		query.Add("out_order_no", *request.OutOrderNo)
	}
	reqUrl.RawQuery = query.Encode()
	httpRequest, err := http.NewRequest(method, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &ReturnOrdersEntity{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type QueryReturnOrderRequest struct {
	OutReturnNo *string `json:"out_return_no,omitempty"`
	SubMchid    *string `json:"sub_mchid,omitempty"`
	OutOrderNo  *string `json:"out_order_no,omitempty"`
}

func (o *QueryReturnOrderRequest) MarshalJSON() ([]byte, error) {
	type Alias QueryReturnOrderRequest
	a := &struct {
		OutReturnNo *string `json:"out_return_no,omitempty"`
		SubMchid    *string `json:"sub_mchid,omitempty"`
		OutOrderNo  *string `json:"out_order_no,omitempty"`
		*Alias
	}{
		OutReturnNo: nil,
		SubMchid:    nil,
		OutOrderNo:  nil,
		Alias:       (*Alias)(o),
	}
	return json.Marshal(a)
}

type ReturnOrdersEntity struct {
	SubMchid    *string                `json:"sub_mchid,omitempty"`
	OrderId     *string                `json:"order_id,omitempty"`
	OutOrderNo  *string                `json:"out_order_no,omitempty"`
	OutReturnNo *string                `json:"out_return_no,omitempty"`
	ReturnId    *string                `json:"return_id,omitempty"`
	ReturnMchid *string                `json:"return_mchid,omitempty"`
	Amount      *int64                 `json:"amount,omitempty"`
	Description *string                `json:"description,omitempty"`
	Result      *ReturnOrderStatus     `json:"result,omitempty"`
	FailReason  *ReturnOrderFailReason `json:"fail_reason,omitempty"`
	CreateTime  *time.Time             `json:"create_time,omitempty"`
	FinishTime  *time.Time             `json:"finish_time,omitempty"`
}

type ReturnOrderStatus string

func (e ReturnOrderStatus) Ptr() *ReturnOrderStatus { return &e }

const (
	RETURNORDERSTATUS_PROCESSING ReturnOrderStatus = "PROCESSING"
	RETURNORDERSTATUS_SUCCESS    ReturnOrderStatus = "SUCCESS"
	RETURNORDERSTATUS_FAILED     ReturnOrderStatus = "FAILED"
)

type ReturnOrderFailReason string

func (e ReturnOrderFailReason) Ptr() *ReturnOrderFailReason { return &e }

const (
	RETURNORDERFAILREASON_ACCOUNT_ABNORMAL       ReturnOrderFailReason = "ACCOUNT_ABNORMAL"
	RETURNORDERFAILREASON_BALANCE_NOT_ENOUGH     ReturnOrderFailReason = "BALANCE_NOT_ENOUGH"
	RETURNORDERFAILREASON_TIME_OUT_CLOSED        ReturnOrderFailReason = "TIME_OUT_CLOSED"
	RETURNORDERFAILREASON_PAYER_ACCOUNT_ABNORMAL ReturnOrderFailReason = "PAYER_ACCOUNT_ABNORMAL"
	RETURNORDERFAILREASON_INVALID_REQUEST        ReturnOrderFailReason = "INVALID_REQUEST"
)
